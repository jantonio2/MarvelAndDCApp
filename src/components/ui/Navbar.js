import React, { useContext } from 'react';
import { NavLink, useHistory } from 'react-router-dom';
import { AuthContext } from './../../auth/AuthContext';
import { types } from './../../types/types';

export const Navbar = () => {

  const {user:{name}, dispatch} = useContext(AuthContext)
  const history = useHistory();
  const handleLogout = () => {
    dispatch({
      type: types.logout
    })
    history.replace('/login')
  }
  
  return (
    <nav className='navbar navbar-expand-lg navbar-dark bg-dark'>
      <div className='container-fluid'>
        <NavLink to='/' className='navbar-brand'>
          ASOCIACIONES
        </NavLink>
        <div className='collapse navbar-collapse' id='navbarText'>
          <ul className='navbar-nav me-auto mb-2 mb-lg-0'>
            <li className='nav-item'>
              <NavLink to='/marvel' className='nav-link'>
                Marvel
              </NavLink>
            </li>
            <li className='nav-item'>
              <NavLink to='/dc' className='nav-link'>
                DC
              </NavLink>
            </li>
            <li className='nav-item'>
              <NavLink to='/search' className='nav-link'>
                Search
              </NavLink>
            </li>
          </ul>
          <ul className='navbar-nav'>
            <li className='nav-item'>
              <span className="nav-link text-info">
                {name}
              </span>
            </li>
            <li className='nav-item'>
              <button 
                className='nav-link btn'
                onClick={handleLogout}
              >
                Logout
              </button>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  )
}
