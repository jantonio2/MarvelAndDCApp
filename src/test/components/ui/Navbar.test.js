import { mount } from 'enzyme'
import { MemoryRouter, Router } from 'react-router-dom'
import { AuthContext } from '../../../auth/AuthContext'
import { Navbar } from '../../../components/ui/Navbar'
import { types } from '../../../types/types'

describe('Pruebas en el componente <Navbar />', () => {

  const historyMock = {
    push: jest.fn(),
    replace: jest.fn(),
    location: {},
    listen: jest.fn(),
    createHref: jest.fn()
  }

  const contextValue = {
    dispatch: jest.fn(),
    user: {
      name: 'Antonio',
      logged: true
    }
  }
  const wrapper = mount(
    <AuthContext.Provider value={contextValue}>
      <MemoryRouter>
        <Router history={historyMock}>
          <Navbar />
        </Router>
      </MemoryRouter>
    </AuthContext.Provider>
  )

  afterEach(() => {
    jest.clearAllMocks()
  })

  test('debe de mostrarse correctamente', () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.find('.text-info').text().trim()).toBe(contextValue.user.name)
  })
  
  test('debe de llamar el logout y usar history', () => {
    wrapper.find('button').prop('onClick')()

    expect(contextValue.dispatch).toHaveBeenCalledWith({type: types.logout})
    expect(historyMock.replace).toHaveBeenCalledWith('/login')
  })

  // Otra forma de hacer lo de arriba

  // const mockHistoryReplace = jest.fn()
 
  // jest.mock('react-router-dom', () => ({
  //     ...jest.requireActual('react-router-dom'),
  //     useHistory: () => ({
  //       replace: mockHistoryReplace
  //     })
  //   })
  // )
  
  // test('debe de llamar el logout y usar history (otra forma)', () => {
  //   wrapper.find('button').prop('onClick')(); // same than .simulate('click')
  
  //   expect(contextValue.dispatch).toHaveBeenCalledWith({type: types.logout});
  //   expect(mockHistoryReplace).toHaveBeenCalledWith('/login');
 
  // })
})
