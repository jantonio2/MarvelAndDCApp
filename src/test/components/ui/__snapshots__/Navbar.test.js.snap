// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <Navbar /> debe de mostrarse correctamente 1`] = `
<nav
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <div
    className="container-fluid"
  >
    <a
      aria-current={null}
      className="navbar-brand"
      onClick={[Function]}
    >
      ASOCIACIONES
    </a>
    <div
      className="collapse navbar-collapse"
      id="navbarText"
    >
      <ul
        className="navbar-nav me-auto mb-2 mb-lg-0"
      >
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-link"
            onClick={[Function]}
          >
            Marvel
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-link"
            onClick={[Function]}
          >
            DC
          </a>
        </li>
        <li
          className="nav-item"
        >
          <a
            aria-current={null}
            className="nav-link"
            onClick={[Function]}
          >
            Search
          </a>
        </li>
      </ul>
      <ul
        className="navbar-nav"
      >
        <li
          className="nav-item"
        >
          <span
            className="nav-link text-info"
          >
            Antonio
          </span>
        </li>
        <li
          className="nav-item"
        >
          <button
            className="nav-link btn"
            onClick={[Function]}
          >
            Logout
          </button>
        </li>
      </ul>
    </div>
  </div>
</nav>
`;
